### PR TITLE
Unhardcode the leeway range of AttackFollow

### LINKED ISSUE #18149: Turreted units move one cell closer than max range when given a target
Not so much a bug, since I've learned this is actually intended behavior.

The rationale is that when chasing, turreted units will constantly go out of range as they stop to fire while the target continues to move. By having them chase 1 cell closer, they never go out of range.

However, this is really detrimental to the RA mod where ranges are key. A lot of combat happens with nothing really moving, and being 1 cell closer is the difference between having your tank in range of 4 rocket soldiers or 15 rocket soldiers. This feature makes it actually a bad play to micro your tanks to target other vehicles, since they'll move closer and get shredded. It also makes ranger micro more difficult.

The workaround (which players did anyway pre-opportunity fire), is to target the enemy vehicle and then issue a move command. With pathfinding potentially being blocked due to friendlies or terrain, this is still the optimal way to chase an enemy target regardless.

### PR DESCRIPTION
**Pull request by abcdefg30**
Allows adressing #18149 via yaml.

~I also set the leeway range of turreted units in RA to 0 as suggested in the issue.~

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Attack/AttackFollow.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Attack/AttackFollow.cs]: The name `LeewayRange` doesn't sound right to me, but I don't have an immediately better suggestion. Maybe something with "Inset" in it?

Also `WDist.FromCells(1)`?

Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Attack/AttackFollow.cs]: Perhaps `LeewayMargin`

**Comment by Orb370** [comment]
It would be good to set this to 0 in TD as well. I don't think it'll have as much an impact on TD but it'd be good to have consistency.

**Comment by pchote** [comment]
At this point, would it be better to change the default to 0 and have it change in all mods?

**Comment by Orb370** [comment]
We have a third mod? :jack_o_lantern: 

But yeah that'd probably be better.

**Comment by abcdefg30** [comment]
Rebased and updated.

File: OpenRA.Mods.Common/Traits/Attack/AttackFollow.cs
Comment (pchote) [reply OpenRA.Mods.Common/Traits/Attack/AttackFollow.cs]: "Leeway" was specifically the part that I didn't like here!

Comment (pchote) [reply OpenRA.Mods.Common/Traits/Attack/AttackFollow.cs]: How about `RangeMargin` instead?

Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/Attack/AttackFollow.cs]: I don't have a strong opinion either way.

Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/Attack/AttackFollow.cs]: Updated.

**Comment by pchote** [comment]
Code changes lgtm, but I can now trivially reproduce the out-of-range issue that the leeway was there to solve.
I find it a bit odd that that people are so quick to dismiss this after splitting so many hairs about the impact of important bugfixes (like the infantry group blocking) on damage dynamics.

Testcase: order one heavy tank to attack a second heavy tank, then order the second tank to start moving as soon as the first one has stopped within range.

1 cell leeway (can fire at full rate):

leeway

no leeway (can't fire at full rate):

noleeway

**Review by PunkPun — DISMISSED** [review]
I played around and haven't encountered any bugs, nor really felt a big impact. I think to know the true impact we need to test it in multiplayer

**Comment by PunkPun** [comment]
> but I can now trivially reproduce the out-of-range issue that the leeway was there to solve.

I don't think that this is an issue. I can't reproduce it in realistic scenarios. Usually the heavy tank just lacks behind as he does more turns than the heavy running away, and in situations were he does catch up, he shoots bursts normally. In very rare cases the slow burst thing happens, but IMO it is quite minor

**Comment by PunkPun** [comment]
Actually I want to dismiss my approval, it crashed [exception-2020-09-09T121209Z.log](https://github.com/OpenRA/OpenRA/files/5195023/exception-2020-09-09T121209Z.log). I played around more after running tests and I'm unsure as to what caused this crash

**Comment by pchote** [comment]
That crash is not related to this PR, and was recently fixed by #18474.

**Review by pchote — APPROVED** [review]
LGTM now that this is only unhardcoding the option without changing anything ingame.

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/be0f58382944526c2703977f12104e101339b138...e644831169aad3db8e9425fa66f8e87f5350b6d5

**Pull request merged**
